extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/config/GrpcConfigBuilder.java

```diff
@@ -0,0 +1,61 @@
+package io.quarkus.grpc.runtime.config;
+
+import static io.vertx.core.net.SocketAddress.inetSocketAddress;
+import static java.lang.Integer.MAX_VALUE;
+import static java.net.InetAddress.getByName;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ConfigBuilder;
+import io.smallrye.config.ConfigSourceContext;
+import io.smallrye.config.ConfigSourceFactory.ConfigurableConfigSourceFactory;
+import io.smallrye.config.SmallRyeConfigBuilder;
+import io.smallrye.config.common.MapBackedConfigSource;
+import io.vertx.core.net.SocketAddress;
+
+public class GrpcConfigBuilder implements ConfigBuilder {
+    @Override
+    public SmallRyeConfigBuilder configBuilder(final SmallRyeConfigBuilder builder) {
+        return builder.withSources(new RandomPortConfigSourceFactory());
+    }
+
+    private static class RandomPortConfigSourceFactory implements ConfigurableConfigSourceFactory<GrpcServerConfig> {
+        @Override
+        public Iterable<ConfigSource> getConfigSources(final ConfigSourceContext context, final GrpcServerConfig config) {
+            Map<String, String> randomPorts = new HashMap<>();
+
+            String host = config.host();
+            if (LaunchMode.current().equals(LaunchMode.TEST)) {
+                if (config.testPort() == 0) {
+                    String port = assignRandomPort(inetSocketAddress(0, host));
+                    randomPorts.put("quarkus.grpc.server.test-port", port);
+                    randomPorts.put("quarkus.grpc.server.port", port);
+                }
+            } else {
+                if (config.port() == 0) {
+                    String port = assignRandomPort(inetSocketAddress(0, host));
+                    randomPorts.put("quarkus.grpc.server.port", port);
+                }
+            }
+            return randomPorts.isEmpty() ? Collections.emptyList()
+                    : List.of(new MapBackedConfigSource("Quarkus GRPC Random Ports", randomPorts, MAX_VALUE - 1000) {
+                    });
+        }
+
+        private static String assignRandomPort(SocketAddress socketAddress) {
```

Comment:
You cannot do that, that's super brittle. 
OS may close things asynchronously, the bind on the same port may fail.

Reply:
Locally and in the CI, things seem to have worked.

What do you suggest to do instead?

Reply:
What I proposed in the issue somewhere:

- a config source that provides the properties for 0, -1, -2... 
- they get the port as you do but add a grace period (configurable)
- the config source does not recompute for a given "random id"

Another tricky thing is that I would like to have a way to keep the current behavior somehow because it got hardened enough.

Reply:
Ok, I didn't notice that we have some special meaning with port numbers here:
https://github.com/quarkusio/quarkus/blob/b41b6edfb6da52217bb10ef14c36a22a5313a55f/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/options/HttpServerOptionsUtils.java#L48-L61

I can implement that behaviour. What threw me off, is that when I was trying the current version and set `0` to `port` and `ssl-port` I would get two different values.

> Another tricky thing is that I would like to have a way to keep the current behavior somehow because it got hardened enough.

Hum... we could add a flag to use new / old behaviour.